### ISSUE TITLE
Upload recursive

### ISSUE THREAD

**Issue by adrianlzt** [bibby/radula#8]
A recursive option for upload could be interesting.
That option should scan directories below the given path and upload with the correct path.

**Comment by bibby** [comment]
Absolutely agree.
The tricky part could be determining "the correct path", as radula is able to make some assumptions when a single file is involved.

From the README:

```
radula up {source} {target}
+--------------+-----------------+-----------------------------------------+
| source       | target          | result                                  |
+==============+=================+=========================================+
| /some/file   | bucket          | bucket/file                             |
+--------------+-----------------+-----------------------------------------+
| /some/file   | bucket/file     | bucket/file                             |
+--------------+-----------------+-----------------------------------------+
| /some/file   | bucket/named    | bucket/named                            |
+--------------+-----------------+-----------------------------------------+
| /some/file   | bucket/named/   | bucket/named/file                       |
+--------------+-----------------+-----------------------------------------+
| /some/f\*    | bucket/named    | bucket/named/file, bucket/named/file2   |
+--------------+-----------------+-----------------------------------------+
| /some/f\*    | bucket/named/   | bucket/named/file, bucket/named/file2   |
+--------------+-----------------+-----------------------------------------+
```

There are cases in that table where radula reuses the source file's original name, so there is precedence for straight *sync*. This has been on my mind for a little while, so I appreciate your mention.

We could certainly use a dir sync. Something along `radula up /path/to/dir bucket/dir`. Issue #9 is a step in this direction, but makes me wonder if we should reupload files if the source is newer somehow, or let that alone.

**Comment by bibby** [comment]
I committed to the `dev` branch, new handling for directories during upload, and globs during download. Together, they effectively constitute recursive handling, but with unique details that are worth noting.

As these are new, I'd like to keep them in the `dev` branch during a testing period so that we can ensure that our expectations are met and (most importantly) all previously existing functionality was not harmed by these changes. You are invited to try the `dev` branch and ticket any issues you may encounter. Remember that the `-n,--dry-run` flag can display changes without making them.

## Uploads

*Assuming a test directory looks like:*
```
dir/dir_file1
dir/dir_file2
file_a
file_b
```

When uploading to ceph-s3, radula can tell if you've selected a directory (directly, or as part of a glob). If spotted, radula will recursively upload the directory's contents using the source path for naming.

`radula -i up '*' bucket` will produce:
```
bucket/dir/dir_file1
bucket/dir/dir_file2
bucket/file_a
bucket/file_b
```

The target can be a path, but a trailing slash is needed!

`radula -p devops -i up '*' bucket/path/` will produce:
```
bucket/path/dir/dir_file1
bucket/path/dir/dir_file2
bucket/path/file_a
bucket/path/file_b
```
If the trailing slash were missing, the two top-level files would have been uploaded to a key called `path`.

The current working directory is important. Uploading a full path will result in a full path on s3, so beware.

`radula -n up  /home/bibby/projects/radula bucket` results in uploads like `bucket/home/abibby/projects/radula/README.rst` (among many others).


As I'm writing this, I've discovered that using a dot for the CWD is not expanded nor ignored. It's uploaded literally, resulting in keys like `bucket/./file_a`. I intended to return to resolve that (by ignoring it).

## Downloads

Downloading files is almost a separate animal. S3 really doesn't have directories per-se, making it difficult to predict what a user may want all of the time. I find myself often wanting a file that is deeply nested on s3 in my current directory. If I wanted to pull down a "directory" as such, I'd expect to have to do something a little different. Downloading does support globs, so at all times, `dl` is expecting to handle multiple files.

So let's presume that s3 has our last uploaded set:
```
bucket/path/dir/dir_file1
bucket/path/dir/dir_file2
bucket/path/file_a
bucket/path/file_b
```

Downloading `path/` or a equivalent glob will download a flat list at the current working directory.

```
radula -n dl bucket/path/
DRY RUN: download-> bucket/path/dir/dir_file1, =>, dir_file1
DRY RUN: download-> bucket/path/dir/dir_file2, =>, dir_file2
DRY RUN: download-> bucket/path/file_a, =>, file_a
DRY RUN: download-> bucket/path/file_b, =>, file_b
```

To achieve a recursive behavior, a new flag `-k,--preserve-keys` has been added that will replicate the key path on the target host, creating directories as needed.

```
radula -n -k dl bucket/path/
DRY RUN: download-> bucket/path/dir/dir_file1, =>, path/dir/dir_file1
DRY RUN: download-> bucket/path/dir/dir_file2, =>, path/dir/dir_file2
DRY RUN: download-> bucket/path/file_a, =>, path/file_a
DRY RUN: download-> bucket/path/file_b, =>, path/file_b
```

Should a final target argument be given to download (ie, a directory to download into), that directory must exist first (otherwise, radula isn't sure what you mean).

*This would be bad:*
```
# 'dl_here' doesn't exist yet
radula -n -k dl bucket/path/ dl_here
DRY RUN: download-> bucket/path/dir/dir_file1, =>, dl_here
DRY RUN: download-> bucket/path/dir/dir_file2, =>, dl_here
DRY RUN: download-> bucket/path/file_a, =>, dl_here
DRY RUN: download-> bucket/path/file_b, =>, dl_here
```

*create it first*
```
mkdir dl_here; radula -n -k dl bucket/path/ dl_here
DRY RUN: download-> bucket/path/dir/dir_file1, =>, dl_here/path/dir/dir_file1
DRY RUN: download-> bucket/path/dir/dir_file2, =>, dl_here/path/dir/dir_file2
DRY RUN: download-> bucket/path/file_a, =>, dl_here/path/file_a
DRY RUN: download-> bucket/path/file_b, =>, dl_here/path/file_b
```

Had I chose a target of `dl_here/` (with a trailing slash), that might have been enough suggestion to radula of my intention, but it does not work.  I'd like to go back and fix that as well.

Hopefully, this is a solid enough start for these features. But, like I say, I'd like to spend a little more time with them before merging. Making sure that existing features still work is the most important thing.

**Comment by adrianlzt** [comment]
Cool. Thank you very much!

**Comment by bibby** [comment]
released as `radula==0.8.0`.  Thanks for all the feedback!
